README
Clipboard Viewer

clipboardviewer [--nogui] [--query]

DESCRIPTION

A GUI to view the main clipboard contents ( text or pic ) or from the command line print the text contents to stdin or create a .png image file in the current directory.

The window is persistant and will be updated regularly.
It can also be made ( the default ) to 'stick' to all workspaces above all over windows.

--nogui - Do not create a window just print the contents to stdin or create a picture in the current folder and print the filename.

--query - Print type of data on clipboard ( image or text ).


ChangeLog
2.2.1
Updated addresses.
Drop down list now reflects the text contents of the clip, truncated to 40 with '...'
2.2
Redone window icon properly.
Compressed manpage.
Updated manpage.

2.1
Increased number of remembered clipboard items to 20.
Added clipboard history for text/images to drop down.

2.0
Now only needs GTK2 no libglade etc.
Window can now be made 'sticky' ie shows above all windows on all work spaces.
Added about box with contact details and GPL3 license.

0.0->1.5
removed gnome centric stuff, now build and runs in xfce ( - the nautilus script )
added man page
added a query command to find out what sort of data is on the clipboard
added a nautilus script to make a text or image clipping in the current directory ala osx
cleaned up makefile
Cleaned up the make file so a Slackware package could be built.
changed links to new website

ClipboardViewer/src/globals.h
/*
 *
 * ©K. D. Hedger. Tue 31 Jul 14:10:54 BST 2018 keithdhedger@gmail.com

 * This file (globals.h) is part of ClipboardViewer.

 * ClipboardViewer is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * at your option) any later version.

 * ClipboardViewer is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with ClipboardViewer.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <gtk/gtk.h>
#include "internet.h"

#ifndef _GLOBALS_
#define _GLOBALS_

#define MAXCLIPS 20
#define MAXDROPTXTLEN 40

struct clips
{
	GdkPixbuf*			image;
	char*				text;
	GtkClipboard*		realclip;
};


extern char*			tempname;
extern GtkClipboard*	mainclipboard;

extern GtkWidget*		scrollBox;
extern GtkWidget*		textBox;
extern GtkWidget*		viewBox;
extern GtkWidget*		bufferBox;
extern GtkWidget*		window;
extern GtkWidget*		notebook;
extern GtkWidget*		scrollPicBox;
extern GtkWidget*		imageBox;
extern GtkWidget*		clipListDrop;

extern int				currentClip;
extern clips			clip[MAXCLIPS];

#endif


ClipboardViewer/src/internet.h
/*
 *
 * ©K. D. Hedger. Tue 28 Jun 10:44:42 BST 2016 kdhedger68713@gmail.com
 
 * Projects is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * at your option) any later version.

 * Projects is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with Projects.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef _INTERNET_
#define _INTERNET_

#define MYEMAIL "keithhedger@keithhedger.darktech.org"
#define MYWEBSITE "http://khapplications.darktech.org/index.html"
#define GLOBALWEBSITE "http://khapplications.darktech.org"
#define ASPELLPAGE GLOBALWEBSITE "/pages/aspellgui/aspelgui.html"
#define MANPAGEPAGE GLOBALWEBSITE "/pages/manpageeditor/help.html"
#define CLIPVIEW GLOBALWEBSITE "/pages/clipviewer/clip.html"

#define COPYRITE "Copyright \xc2\xa9 2013-2016 K.D.Hedger"

#ifndef _USEQT5_
extern const char	*authors[];
#else
extern const char*	authors;
#endif
#endif


ClipboardViewer/src/globals.cpp
/*
 *
 * ©K. D. Hedger. Tue 31 Jul 14:10:47 BST 2018 keithdhedger@gmail.com

 * This file (globals.cpp) is part of ClipboardViewer.

 * ClipboardViewer is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * at your option) any later version.

 * ClipboardViewer is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with ClipboardViewer.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "globals.h"

char*			tempname;
GtkClipboard*	mainclipboard;

GtkWidget*		scrollBox=NULL;
GtkWidget*		textBox=NULL;
GtkWidget*		viewBox=NULL;
GtkWidget*		bufferBox=NULL;
GtkWidget*		window=NULL;
GtkWidget*		notebook=NULL;
GtkWidget*		scrollPicBox=NULL;
GtkWidget*		imageBox=NULL;
GtkWidget*		clipListDrop=NULL;

int				currentClip=0;
clips			clip[MAXCLIPS];


ClipboardViewer/src/main.cpp
/*
 *
 * ©K. D. Hedger. Tue 31 Jul 14:11:05 BST 2018 keithdhedger@gmail.com

 * This file (main.cpp) is part of ClipboardViewer.

 * ClipboardViewer is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * at your option) any later version.

 * ClipboardViewer is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with ClipboardViewer.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <glib.h>

#include "globals.h"
#include "config.h"

bool		manual=false;

char* truncateText(char* txt)
{
	char*	retstr;
	char*	formatstr;
	int		side=(MAXDROPTXTLEN-5)/2;

	if(strlen(txt)<=MAXDROPTXTLEN)
		{
			retstr=strdup(txt);
		}
	else
		{
			asprintf(&formatstr,"%%.%is ... %%s",side);
			asprintf(&retstr,formatstr,txt,&txt[strlen(txt)-side]);
			free(formatstr);
		}

	for(unsigned j=0;j<strlen(retstr);j++)
	 if(retstr[j]=='\n')
		 	retstr[j]='.';

	return(retstr);
}

void setCurrentClip(void)
{
	currentClip++;
	if(currentClip==MAXCLIPS)
		currentClip=0;
}

void showContents(int clipnum)
{
	if(clip[clipnum].text!=NULL)
		{
			gtk_text_buffer_set_text((GtkTextBuffer*)bufferBox,clip[clipnum].text,-1);
			gtk_notebook_set_current_page((GtkNotebook*)notebook,0);
			return;
		}

	if(clip[clipnum].image!=NULL)
		{
			gtk_image_set_from_pixbuf((GtkImage*)imageBox,clip[clipnum].image);
			gtk_notebook_set_current_page((GtkNotebook*)notebook,1);
			return;
		}
}

void clipChanged(GtkClipboard* clipboard,gpointer user_data)
{
	char*	droptext=NULL;

	if (manual==true)
		{
			manual=false;
			return;
		}
	
	if (gtk_clipboard_wait_is_text_available(mainclipboard)==true)
		{
			setCurrentClip();
			if(clip[currentClip].text != NULL)
				free(clip[currentClip].text);
			clip[currentClip].text=gtk_clipboard_wait_for_text(clipboard);

			if(clip[currentClip].image != NULL)
				{
					g_object_unref(clip[currentClip].image);
					clip[currentClip].image=NULL;
				}
			gtk_notebook_set_current_page((GtkNotebook*)notebook,0);

			gtk_combo_box_text_remove((GtkComboBoxText*)clipListDrop,currentClip);
			droptext=truncateText(clip[currentClip].text);
			gtk_combo_box_text_insert_text((GtkComboBoxText*)clipListDrop,currentClip,droptext);
			free(droptext);

			gtk_combo_box_set_active((GtkComboBox*)clipListDrop,currentClip);
			showContents(currentClip);
		}

	if (gtk_clipboard_wait_is_image_available(mainclipboard)==true)
		{

			setCurrentClip();
			if(clip[currentClip].text != NULL)
				{
					free(clip[currentClip].text);
					clip[currentClip].text=NULL;
				}
			if(clip[currentClip].image != NULL)
				g_object_unref(clip[currentClip].image);

			clip[currentClip].image=gtk_clipboard_wait_for_image(clipboard);
			gtk_notebook_set_current_page((GtkNotebook*)notebook,0);
			gtk_combo_box_set_active((GtkComboBox*)clipListDrop,currentClip);
			showContents(currentClip);
		}


	clip[currentClip].realclip=gtk_clipboard_get(GDK_SELECTION_CLIPBOARD);
	gtk_clipboard_store(clip[currentClip].realclip);
}

void setClip(GtkWidget* widget,gpointer data)
{
	int clipnum=gtk_combo_box_get_active((GtkComboBox*)widget);

	showContents(clipnum);
	if((clip[clipnum].text !=NULL) && (clip[clipnum].realclip!=NULL))
		{
			manual=true;
			gtk_clipboard_set_text(mainclipboard,clip[clipnum].text,-1);
			showContents(clipnum);
		}
	if((clip[clipnum].image !=NULL) && (clip[clipnum].realclip!=NULL))
		{
			manual=true;
			gtk_clipboard_set_image(mainclipboard,clip[clipnum].image);
			showContents(clipnum);
		}
	return;
}

void doShutdown(GtkButton *button, gpointer window_ptr)
{
	gtk_main_quit();
	exit(0);
}

void doAbout(GtkWidget* widget,gpointer data)
{
	const char	copyright[]=COPYRITE;
	char*		license=NULL;
	char*		doc=NULL;
	FILE*		fd=NULL;
	long		fsize=0;

	asprintf(&doc,"%s/docs/gpl-3.0.txt",DATADIR);
	fd=fopen(doc,"r");

	if(fd!=NULL)
		{
			fseek(fd,0,SEEK_END);
			fsize=ftell(fd);
			fseek(fd,0,SEEK_SET);
			license=(char*)malloc(fsize+1);
			fread(license,fsize,1,fd);
			fclose(fd);
			license[fsize]=0;
		}

	gtk_show_about_dialog(NULL,"authors",authors,"copyright",copyright,"version",VERSION,"website",CLIPVIEW,"website-label","Clipboard Viewer Page","program-name","ClipboardViewer","logo-icon-name","ClipboardViewer","license",license,NULL); 
}

void doSticky(GtkWidget* widget,gpointer data)
{
	if(gtk_toggle_button_get_active((GtkToggleButton*)widget)==true)
		{
			gtk_window_stick(GTK_WINDOW(window));
			gtk_button_set_label((GtkButton*)widget,"Un-Stick");
			gtk_window_set_keep_above((GtkWindow*)window,true);
		}
	else
		{
			gtk_window_unstick(GTK_WINDOW(window));		
			gtk_button_set_label((GtkButton*)widget,"Stick");
			gtk_window_set_keep_above((GtkWindow*)window,false);
		}

}

void buildMainGui(void)
{
	GtkWidget*	vbox;
	GtkWidget*	hbox;
	GtkWidget*	button;

	window=gtk_window_new(GTK_WINDOW_TOPLEVEL);
	gtk_window_set_title((GtkWindow*)window,"Clipboard");
	gtk_window_set_default_size((GtkWindow*)window,320,60);
	vbox=gtk_vbox_new(false,8);

	notebook=gtk_notebook_new();
	gtk_notebook_set_show_tabs((GtkNotebook*)notebook,false);

//text clip
	scrollBox=gtk_scrolled_window_new(NULL, NULL);
	gtk_scrolled_window_set_policy(GTK_SCROLLED_WINDOW(scrollBox),GTK_POLICY_AUTOMATIC,GTK_POLICY_AUTOMATIC);
	bufferBox=(GtkWidget*)gtk_text_buffer_new(NULL);
	viewBox=(GtkWidget*)gtk_text_view_new_with_buffer((GtkTextBuffer*)bufferBox);
	gtk_text_view_set_wrap_mode((GtkTextView*)viewBox,GTK_WRAP_WORD);
	gtk_container_add(GTK_CONTAINER(scrollBox),(GtkWidget*)viewBox);

	gtk_notebook_append_page((GtkNotebook*)notebook,scrollBox,NULL);

//image clip
	scrollPicBox=gtk_scrolled_window_new(NULL, NULL);
	gtk_scrolled_window_set_policy(GTK_SCROLLED_WINDOW(scrollPicBox),GTK_POLICY_AUTOMATIC,GTK_POLICY_AUTOMATIC);
	imageBox=gtk_image_new();
	gtk_scrolled_window_add_with_viewport((GtkScrolledWindow*)scrollPicBox,imageBox);

	gtk_notebook_append_page((GtkNotebook*)notebook,scrollPicBox,NULL);

	gtk_container_add(GTK_CONTAINER(vbox),(GtkWidget*)notebook);

//buttons
	hbox=gtk_hbox_new(false,8);

	button=gtk_button_new_from_stock(GTK_STOCK_ABOUT);
	g_signal_connect_after(G_OBJECT(button),"clicked",G_CALLBACK(doAbout),NULL);
	gtk_box_pack_start(GTK_BOX(hbox),button,false,false,4);

	button=gtk_toggle_button_new_with_label("Un-Stick");
	gtk_toggle_button_set_active((GtkToggleButton*)button,true);
	g_signal_connect(G_OBJECT(button),"clicked",G_CALLBACK(doSticky),NULL);
	gtk_box_pack_start(GTK_BOX(hbox),button,true,true,2);

//combo box
	char	tbuff[64]={0};
	clipListDrop=gtk_combo_box_text_new();
	for(int j=0;j<MAXCLIPS;j++)
		{
			sprintf(tbuff,"Clip No. %i",j);
			gtk_combo_box_text_append_text((GtkComboBoxText*)clipListDrop,tbuff);
			clip[j].text=NULL;
			clip[j].image=NULL;
			clip[j].realclip=NULL;
		}
	gtk_combo_box_set_active((GtkComboBox*)clipListDrop,0);
	g_signal_connect(G_OBJECT(clipListDrop),"changed",G_CALLBACK(setClip),NULL);
	gtk_box_pack_start(GTK_BOX(hbox),clipListDrop,true,true,2);

	button=gtk_button_new_from_stock(GTK_STOCK_QUIT);
	g_signal_connect(G_OBJECT(button),"clicked",G_CALLBACK(doShutdown),NULL);
	gtk_box_pack_start(GTK_BOX(hbox),button,false,false,2);
	gtk_box_pack_start(GTK_BOX(vbox),(GtkWidget*)hbox,false,false,2);

	g_signal_connect(G_OBJECT(window),"delete-event",G_CALLBACK(doShutdown),NULL);
	gtk_container_add(GTK_CONTAINER(window),(GtkWidget*)vbox);
}

int main(int argc, char **argv)
{
	gtk_init(&argc,&argv);

	mainclipboard=gtk_clipboard_get(GDK_SELECTION_CLIPBOARD);
	g_signal_connect(G_OBJECT(mainclipboard),"owner-change",G_CALLBACK(clipChanged),NULL);

	tempname=tempnam("./","image");
	strcat(tempname,".png");

	if (argc>1 && g_ascii_strcasecmp(argv[1],"--nogui")==0)
		{
			GdkPixbuf	*image=gtk_clipboard_wait_for_image(mainclipboard);
		
			if (image!=NULL)
				{
					gdk_pixbuf_save(image,tempname, "png",NULL,NULL, NULL, NULL);
					g_object_unref((gpointer) image);
					printf("%s",tempname);
					free(tempname);
				}
			else
				{
					gchar	*clipText=gtk_clipboard_wait_for_text(mainclipboard);

				if (clipText!=NULL)
					{
						printf("%s",clipText);
						free(clipText);
					}
				}
			return 0;
		}

	if (argc>1 && g_ascii_strcasecmp(argv[1],"--query")==0)
		{
			GdkPixbuf	*image=gtk_clipboard_wait_for_image(mainclipboard);
			if (image!=NULL)
				{
					printf("image\n");
					g_object_unref((gpointer) image);
					return 0;
				}
			gchar	*clipText=gtk_clipboard_wait_for_text(mainclipboard);
			if (clipText!=NULL)
				{
					printf("text\n");
					g_free(clipText);
					return 0;
				}
		}

	if (argc>1)
		{
			printf("clipboardviewer %s\n%s: invalid option\n",VERSION,argv[1]);
			printf("Usage:	clipboardviewer [--nogui] [--query]\n");
			return 1;
		}

	buildMainGui();
	gtk_window_stick(GTK_WINDOW(window));
	gtk_window_set_keep_above((GtkWindow*)window,true);
	clipChanged(mainclipboard,NULL);
	gtk_window_set_default_icon_name(PACKAGE);
	gtk_widget_show_all(window);
	
	gtk_main();
}
